File: go/1536.go
package main

func isValid(grid [][]int) bool {
	for i := 0; i < len(grid); i++ {
		for j := i + 1; j < len(grid); j++ {
			if grid[i][j] != 0 {
				return false
			}
		}
	}

	return true
}

func minSwaps(grid [][]int) int {
	result := 0
	maxRight := []int{}
	for _, row := range grid {
		idx := len(row) - 1

		for idx >= 0 {
			if row[idx] == 1 {
				maxRight = append(maxRight, idx)
				break
			}
		}
	}

	for i := 0; i < len(grid); i++ {
		k := -1
		for j := i; j < len(grid); j++ {
			if maxRight[j] <= i {
				result += j - i
				k = j
				break
			}
		}

		if k != -1 {
			for j := k; j > i; j-- {
				maxRight[j], maxRight[j-1] = maxRight[j-1], maxRight[j]

			}
		} else {
			return -1
		}
	}

	return result
}

File: zig/1536.zig
const std = @import("std");

fn solution(alloc: std.mem.Allocator, grid: [][]const i32) !i32 {
    const n = grid.len;
    var result: i32 = 0;

    var max_right = try alloc.alloc(usize, n);
    defer alloc.free(max_right);

    for (grid, 0..) |row, i| {
        var last_one_idx: usize = 0;
        var j: usize = row.len;
        while (j > 0) {
            j -= 1;
            if (row[j] == 1) {
                last_one_idx = j;
                break;
            }
        }
        max_right[i] = last_one_idx;
    }

    for (0..n) |i| {
        var pos: ?usize = null;

        for (i..n) |j| {
            if (max_right[j] <= i) {
                pos = j;
                break;
            }
        }

        if (pos) |found_idx| {
            result += @intCast(found_idx - i);

            var k = found_idx;
            while (k > i) : (k -= 1) {
                std.mem.swap(usize, &max_right[k], &max_right[k - 1]);
            }
        } else {
            // Return -1 as an i32 to signal failure
            return -1;
        }
    }

    return result;
}
